resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/Client.scala

```diff
@@ -0,0 +1,234 @@
+    val driverJavaOptsEnvs: Seq[EnvVar] = resolvedDriverJavaOpts.zipWithIndex.map {
+      case (option, index) =>
+        new EnvVarBuilder()
+          .withName(s"$ENV_JAVA_OPT_PREFIX$index")
+          .withValue(option)
+          .build()
+    }
+
+    val resolvedDriverContainer = new ContainerBuilder(currentDriverSpec.driverContainer)
+      .addAllToEnv(driverJavaOptsEnvs.asJava)
+      .build()
+    val resolvedDriverPod = new PodBuilder(currentDriverSpec.driverPod)
+      .editSpec()
+        .addToContainers(resolvedDriverContainer)
+        .endSpec()
+      .build()
+
+    Utils.tryWithResource(
+        kubernetesClient
+          .pods()
+          .withName(resolvedDriverPod.getMetadata.getName)
+          .watch(loggingPodStatusWatcher)) { _ =>
+      val createdDriverPod = kubernetesClient.pods().create(resolvedDriverPod)
+      try {
+        if (currentDriverSpec.otherKubernetesResources.nonEmpty) {
+          val otherKubernetesResources = currentDriverSpec.otherKubernetesResources
+          addDriverOwnerReference(createdDriverPod, otherKubernetesResources)
+          kubernetesClient.resourceList(otherKubernetesResources: _*).createOrReplace()
+        }
+      } catch {
+        case e: Throwable =>
+          kubernetesClient.pods().delete(createdDriverPod)
+          throw e
+      }
+
+      if (waitForAppCompletion) {
+        logInfo(s"Waiting for application $appName to finish...")
+        loggingPodStatusWatcher.awaitCompletion()
+        logInfo(s"Application $appName finished.")
+      } else {
+        logInfo(s"Deployed Spark application $appName into Kubernetes.")
+      }
+    }
+  }
+
+  // Add a OwnerReference to the given resources making the driver pod an owner of them so when
+  // the driver pod is deleted, the resources are garbage collected.
+  private def addDriverOwnerReference(driverPod: Pod, resources: Seq[HasMetadata]): Unit = {
+    val driverPodOwnerReference = new OwnerReferenceBuilder()
+      .withName(driverPod.getMetadata.getName)
+      .withApiVersion(driverPod.getApiVersion)
+      .withUid(driverPod.getMetadata.getUid)
+      .withKind(driverPod.getKind)
+      .withController(true)
+      .build()
+    resources.foreach { resource =>
+      val originalMetadata = resource.getMetadata
+      originalMetadata.setOwnerReferences(Collections.singletonList(driverPodOwnerReference))
+    }
+  }
+}
+
+/**
+ * Main class and entry point of application submission in KUBERNETES mode.
+ */
+private[spark] object Client extends SparkApplication {
+
+  override def start(args: Array[String], conf: SparkConf): Unit = {
+    val parsedArguments = ClientArguments.fromCommandLineArgs(args)
+    run(parsedArguments, conf)
+  }
+
+  private def run(clientArguments: ClientArguments, sparkConf: SparkConf): Unit = {
+    val namespace = sparkConf.get(KUBERNETES_NAMESPACE)
+    val kubernetesAppId = s"spark-${UUID.randomUUID().toString.replaceAll("-", "")}"
```

Comment:
Is it an application Id for Spark k8s application? Does K8s have a convention for app id? Normally like YARN, it is timestamp based, which is easy for sorting, retrieving, I'm not sure how k8s treat this?

From my understanding of the code, seems like k8s doesn't have app id related concept, so it is up to Spark to define here? Is it posssible to change to timestamp based like standalone, local and yarn deploy?

Reply:
> Does K8s have a convention for app id?

K8s does not. It does have the notion of UUIDs for individual objects, but those are associated with each pod, and not common to an entire application. For representation, we will be introducing a [CRD](https://kubernetes.io/docs/concepts/api-extension/custom-resources/) in the future. That will be how we track individual spark apps in future (post 2.3). That CRD should be able to accommodate all the details, such as timestamp, and status - so, keeping the appID (and driver pod names) collision free using UUIDs made sense.

Reply:
Kubernetes is different from YARN in the sense that the API server expects one to provide their own names for things that are unique in their namespace. However we're coming up with the unique object names for the user since a single spark-submit command generates multiple kinds of objects (secrets, service, pod) that may end up being difficult for the user to reason about in terms of uniqueness.

Reply:
Thanks for the explanation. 

My hunch is that if we can make this app-id more meaningful, like timestamp-xxx, which will be easier for user to distinguish. Using UUID though achieves uniqueness, seems not so intuitive.

Reply:
The app id is actually not meant for users to see and use so being readable is not important. It's used as a label selector to group k8s API resources belonging to the same application. The resource name prefix, on the other hand, does have the application name in it so users know which application the resources belong to.